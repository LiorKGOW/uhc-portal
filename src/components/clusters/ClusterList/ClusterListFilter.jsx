import React from 'react';
import { Filter, FormControl } from 'patternfly-react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { onListFilterSet } from '../../../redux/actions/viewOptionsActions';
import { viewConstants } from '../../../redux/constants';


class ClusterListFilter extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      // The current input value is in the local state, while the currently
      // set filter is in the redux state.
      // This is done to allow some delay between the user's input and
      // the actual filtering, to give them time to finish typing.
      currentValue: '',
    };
    this.updateFilter = this.updateFilter.bind(this);
    this.inputTimeoutID = null;
  }

  componentDidMount() {
    const { currentFilter } = this.props;
    if (currentFilter) {
      this.setState({ currentValue: currentFilter });
    }
  }

  updateFilter() {
    const { setFilter } = this.props;
    const { currentValue } = this.state;
    setFilter(currentValue);
  }

  updateCurrentValue(event) {
    if (this.inputTimeoutID !== null) {
      clearTimeout(this.inputTimeoutID);
    }
    this.setState({ currentValue: event.target.value });
    this.inputTimeoutID = setTimeout(this.updateFilter, 300);
  }

  render() {
    const { currentValue } = this.state;
    return (
      <Filter className="pull-right cluster-list-top">
        <FormControl
          type="text"
          value={currentValue}
          placeholder="Filter by name or ID..."
          onChange={e => this.updateCurrentValue(e)}
        />
      </Filter>);
  }
}

ClusterListFilter.propTypes = {
  currentFilter: PropTypes.string.isRequired,
  setFilter: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  currentFilter: state.viewOptions[viewConstants.CLUSTERS_VIEW].filter,
});

const mapDispatchToProps = {
  setFilter: filter => onListFilterSet(filter, viewConstants.CLUSTERS_VIEW),
};

export default connect(mapStateToProps, mapDispatchToProps)(ClusterListFilter);
