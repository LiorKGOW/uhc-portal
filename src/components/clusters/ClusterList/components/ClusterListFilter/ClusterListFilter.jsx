import React from 'react';
import { Filter, FormControl } from 'patternfly-react';
import PropTypes from 'prop-types';

class ClusterListFilter extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      // The current input value is in the local state, while the currently
      // set filter is in the redux state.
      // This is done to allow some delay between the user's input and
      // the actual filtering, to give them time to finish typing.
      currentValue: '',
    };
    this.updateFilter = this.updateFilter.bind(this);
    this.inputTimeoutID = null;
  }

  componentDidMount() {
    const { currentFilter } = this.props;
    if (currentFilter) {
      this.setState({ currentValue: currentFilter });
    }
  }

  updateFilter() {
    const { setFilter } = this.props;
    const { currentValue } = this.state;
    setFilter(currentValue);
  }

  updateCurrentValue(event) {
    if (this.inputTimeoutID !== null) {
      clearTimeout(this.inputTimeoutID);
    }
    this.setState({ currentValue: event.target.value });
    this.inputTimeoutID = setTimeout(this.updateFilter, 300);
  }

  render() {
    const { currentValue } = this.state;
    return (
      <Filter className="pull-right cluster-list-top">
        <FormControl
          type="text"
          value={currentValue}
          placeholder="Filter by name or ID..."
          onChange={e => this.updateCurrentValue(e)}
        />
      </Filter>);
  }
}

ClusterListFilter.propTypes = {
  currentFilter: PropTypes.string.isRequired,
  setFilter: PropTypes.func.isRequired,
};

export default ClusterListFilter;
