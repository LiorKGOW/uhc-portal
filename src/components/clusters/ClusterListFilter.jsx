import React from 'react';
import { Filter, FormControl } from 'patternfly-react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import * as viewActions from '../../redux/actions/viewOptionsActions';
import { viewConstants } from '../../redux/constants';


class ClusterListFilter extends React.Component {
  state = {
    // The current filter value is only relevant internally, so it shouldn't be in redux
    currentValue: '',
  };


  updateCurrentValue(event) {
    const { setFilter } = this.props;
    this.setState({ currentValue: event.target.value });
    setFilter(`name like '%${event.target.value}%'`);
  }


  render() {
    const { currentValue } = this.state;
    return (
      <Filter>
        <FormControl
          type="text"
          value={currentValue}
          placeholder="Filter by name..."
          onChange={e => this.updateCurrentValue(e)}
        />
      </Filter>);
  }
}

ClusterListFilter.propTypes = {
  setFilter: PropTypes.func.isRequired,
};


const mapDispatchToProps = {
  setFilter: filter => viewActions.onListFilterSet(filter, viewConstants.CLUSTERS_VIEW),
};

export default connect(null, mapDispatchToProps)(ClusterListFilter);
