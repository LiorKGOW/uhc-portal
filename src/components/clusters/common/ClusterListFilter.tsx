import React from 'react';
import { useDispatch } from 'react-redux';

import { TextInput } from '@patternfly/react-core';

import { onListFilterSet } from '~/redux/actions/viewOptionsActions';
import { useGlobalState } from '~/redux/hooks';

type ClusterListFilterProps = {
  view: string;
  isDisabled?: boolean;
};

const ClusterListFilter = ({ view, isDisabled }: ClusterListFilterProps) => {
  const selectRef = React.useRef<HTMLInputElement>(null);
  const dispatch = useDispatch();

  // The current input value is in the local state, while the currently
  // set filter is in the redux state.
  // This is done to allow some delay between the user's input and
  // the actual filtering, to give them time to finish typing.
  const [currentValue, setCurrentValue] = React.useState<
    | string
    | { description?: string; loggedBy?: string; timestampFrom?: string; timestampTo?: string }
  >('');
  const [inputTimeoutID, setInputTimeoutID] = React.useState<any>(undefined);
  const currentFilter = useGlobalState((state) => state.viewOptions[view].filter);

  const updateCurrentValue = (value: string) => {
    if (inputTimeoutID !== undefined) {
      clearTimeout(inputTimeoutID);
    }
    setInputTimeoutID(setTimeout(() => dispatch(onListFilterSet(value, view)), 300));
    setCurrentValue(value);
  };

  // setting focus on the TextInput after the value is updated
  React.useLayoutEffect(() => {
    if (!isDisabled && currentFilter) selectRef.current?.focus();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isDisabled, currentFilter]);

  React.useEffect(() => {
    if (currentFilter) {
      setCurrentValue(currentFilter);
    }

    return () => {
      setCurrentValue('');
      dispatch(onListFilterSet('', view));
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <TextInput
      type="text"
      aria-label="Filter"
      className="cluster-list-filter"
      ref={selectRef}
      value={currentValue as any}
      placeholder="Filter by name or ID..."
      data-testid="filterInputClusterList"
      onChange={(_event, inputVal: string) => updateCurrentValue(inputVal)}
      isDisabled={isDisabled}
    />
  );
};

export default ClusterListFilter;
