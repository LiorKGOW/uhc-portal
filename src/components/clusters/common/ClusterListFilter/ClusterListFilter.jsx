import React from 'react';
import { TextInput } from '@patternfly/react-core';
import PropTypes from 'prop-types';

class ClusterListFilter extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      // The current input value is in the local state, while the currently
      // set filter is in the redux state.
      // This is done to allow some delay between the user's input and
      // the actual filtering, to give them time to finish typing.
      currentValue: '',
    };
    this.updateFilter = this.updateFilter.bind(this);
    this.inputTimeoutID = null;
  }

  componentDidMount() {
    const { currentFilter } = this.props;
    if (currentFilter) {
      this.setState({ currentValue: currentFilter });
    }
  }

  updateFilter() {
    const { setFilter } = this.props;
    const { currentValue } = this.state;
    setFilter(currentValue);
  }

  updateCurrentValue(value) {
    if (this.inputTimeoutID !== null) {
      clearTimeout(this.inputTimeoutID);
    }
    this.setState({ currentValue: value });
    this.inputTimeoutID = setTimeout(this.updateFilter, 300);
  }

  render() {
    const { currentValue } = this.state;
    const { isDisabled } = this.props;
    return (
      <TextInput
        type="text"
        aria-label="Filter"
        className="cluster-list-filter"
        value={currentValue}
        placeholder="Filter by name or ID..."
        onChange={value => this.updateCurrentValue(value)}
        isDisabled={isDisabled}
      />
    );
  }
}

ClusterListFilter.propTypes = {
  currentFilter: PropTypes.string.isRequired,
  setFilter: PropTypes.func.isRequired,
  isDisabled: PropTypes.bool,
};

export default ClusterListFilter;
