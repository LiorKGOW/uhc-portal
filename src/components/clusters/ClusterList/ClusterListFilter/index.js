
import { connect } from 'react-redux';
import { onListFilterSet } from '../../../../redux/actions/viewOptionsActions';
import { viewConstants } from '../../../../redux/constants';
import ClusterListFilter from './ClusterListFilter';

const mapStateToProps = state => ({
  currentFilter: state.viewOptions[viewConstants.CLUSTERS_VIEW].filter,
});

const mapDispatchToProps = {
  setFilter: filter => onListFilterSet(filter, viewConstants.CLUSTERS_VIEW),
};

export default connect(mapStateToProps, mapDispatchToProps)(ClusterListFilter);
