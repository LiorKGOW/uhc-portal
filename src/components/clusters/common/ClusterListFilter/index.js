
import { connect } from 'react-redux';
import { onListFilterSet } from '../../../../redux/actions/viewOptionsActions';
import ClusterListFilter from './ClusterListFilter';

const mapStateToProps = (state, ownProps) => ({
  currentFilter: state.viewOptions[ownProps.view].filter,
});

const mapDispatchToProps = (dispatch, ownProps) => ({
  setFilter: filter => dispatch(onListFilterSet(filter, ownProps.view)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ClusterListFilter);
