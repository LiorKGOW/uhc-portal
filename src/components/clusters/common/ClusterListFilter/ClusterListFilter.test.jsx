import React from 'react';
import { shallow } from 'enzyme';
import { TextInput } from '@patternfly/react-core';

import ClusterListFilter from './ClusterListFilter';

jest.useFakeTimers('legacy'); // TODO 'modern'

describe('<ClusterListFilter />', () => {
  let setFilter;
  let wrapper;
  let wrapperWithPrefilledText;

  beforeEach(() => {
    setFilter = jest.fn();
    wrapper = shallow(<ClusterListFilter
      setFilter={setFilter}
      currentFilter=""
    />);

    wrapperWithPrefilledText = shallow(<ClusterListFilter
      setFilter={setFilter}
      currentFilter="hello"
    />);
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
    expect(wrapperWithPrefilledText).toMatchSnapshot();
  });

  it('sets up a timeout properly', () => {
    wrapper.find(TextInput).simulate('change', 'hello');
    expect(setTimeout).toBeCalled();
    expect(setTimeout).toHaveBeenLastCalledWith(expect.any(Function), 300);
  });

  it('calls setFilter after timeout has passed', () => {
    wrapper.find(TextInput).simulate('change', 'world');
    jest.runOnlyPendingTimers();
    expect(setFilter).toHaveBeenCalledWith('world');
  });

  it('calls setFilter only when the user stops typing', () => {
    wrapper.find(TextInput).simulate('change', 'a');
    wrapper.find(TextInput).simulate('change', 'abc');
    jest.runOnlyPendingTimers();
    expect(setFilter).not.toHaveBeenCalledWith('a');
    expect(setFilter).toHaveBeenCalledWith('abc');
  });
});
