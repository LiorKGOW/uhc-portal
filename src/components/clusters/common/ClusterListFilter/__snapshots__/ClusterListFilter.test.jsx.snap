// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClusterListFilter /> renders correctly 1`] = `
<TextInput
  aria-label="Filter"
  className="cluster-list-filter"
  isDisabled={false}
  isReadOnly={false}
  isRequired={false}
  isValid={true}
  onChange={[Function]}
  placeholder="Filter by name or ID..."
  type="text"
  value=""
/>
`;

exports[`<ClusterListFilter /> renders correctly 2`] = `
<TextInput
  aria-label="Filter"
  className="cluster-list-filter"
  isDisabled={false}
  isReadOnly={false}
  isRequired={false}
  isValid={true}
  onChange={[Function]}
  placeholder="Filter by name or ID..."
  type="text"
  value="hello"
/>
`;
